Don't allow installing packages without ARCH or OS

ffesti opened the pull request:
We had an exception for public key packages in rpmte. With those now being handled in the keystore without going through the transaction machinery this is no longer needed. This also prevents people from just contrsucting their own pubkey packages and install them.

Resolves: #3344

pmatilai commented:
This wants tests too. Adding elements can be easily tested by creating a partial header with the python bindings and tossing it at ts.addInstall(). Erase can be tested on the database in tests/data/misc, there's a gpg-pubkey in there, but what the exact outcome of that should be, lets decide in the ticket first.

ffesti commented:
OK, added all kind of error handling and check - even if they don't quite fit into the #3344 umbrella.

Test are still missing as it is not quite as easy to get a "properly" broken v4 header constructed in the Python API - now that v3 headers without RPMTAG_HEADERIMMUTABLE are no longer allowed.

ffesti commented:
I wonder if I can just take a proper header from on of our test packages and remove tags to test the error handling.

ffesti commented:
Turns out the answer is "yes".

pmatilai commented on lib/rpmte.cc:
```diff
@@ -151,17 +151,37 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
+	    if (p->arch == NULL)
+		rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), "arch");
+	    if (p->os == NULL)
+		rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), "os");
```
If you want to add an error message, and one central "malformed package" style thing at the end of addTE() on error, that'll cover all the existing sanity checks too. There are a hundred thousand ways things can go wrong and we can't have specific messages for each.

pmatilai commented on lib/rpmte.cc:
```diff
@@ -151,17 +151,37 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
+	      {p->release, "release"}}) {
+	if (value == NULL) {
+	    rpmlog(RPMLOG_ERR, _("Package is missing %s\n"), name.c_str());
+	    goto exit;
```
Nah, just leave this like it was, and with a central error message at the end if you like. These are not "packager did something wrong" situations where the user could fix something, it's just a malformed package.

ffesti replied:
Hey, I was sooo proud of getting this loop to actually work...

pmatilai commented on lib/rpmte.cc:
```diff
@@ -151,17 +151,37 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
-    if (!rstreq(p->name, "gpg-pubkey") && (p->arch == NULL || p->os == NULL))
+    if (p->arch == NULL || p->os == NULL) {
+	if (p->type == TR_REMOVED && rstreq(p->name, "gpg-pubkey")) {
+	    rpmlog(RPMLOG_WARNING, "Erasing gpg-pubkey packages is deprecated! Use rpmkeys.\n");
```
We could take the opportunity to be helpful and give the actual rpmkeys command to use here. 
But don't use exclamation marks in messages, it just doesn't belong - the computer is not yelling at you. Certainly not in a harmless situation like this.

pmatilai commented:
> I wonder if I can just take a proper header from on of our test packages and remove tags to test the error handling.

It's an interesting thing. Deleting the name tag like in the test doesn't actually delete it in the header because it's in the immutable region, it just replaces what you see from the "outside", this is basically how file relocation works. But it's a bit weird that you can replace a name, version etc and the header will still pass signature checking inside rpmdb. You can only do that from the API, but it's another strange loophole we probably should close one of these days (obviously way out of scope here)

ffesti commented:
OK, adjusted the messages and reverted to the old check for missing tags.

Last patch still doesn't have a test case. But we just made it impossible to even create such a scenario...

pmatilai commented on lib/rpmte.cc:
```diff
@@ -159,9 +159,18 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
-    if (!rstreq(p->name, "gpg-pubkey") && (p->arch == NULL || p->os == NULL))
+    if (p->arch == NULL || p->os == NULL) {
+	if (p->type == TR_REMOVED && rstreq(p->name, "gpg-pubkey")) {
+	    rpmlog(RPMLOG_WARNING, "erasing gpg-pubkey packages is deprecated. use rpmkeys --delete\n");
```
If you use a period then you'd need to continue the next sentence with a capital letter. Which then requires that for the first sentence too. Use a comma, maybe? And like I already suggested, we could just as easily print out the actual rpmkeys command to use for deleting this particular key, so I think we should.

pmatilai commented on lib/rpmte.cc:
```diff
@@ -159,9 +159,18 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
+    }
+
+    if (p->type != TR_REMOVED && rstreq(p->name, "gpg-pubkey")) {
+	rpmlog(RPMLOG_ERR, "installing gpg-pubkey packages is not allowed. use rpmkeys --import instead\n");
```
Same issue with the period, comma, capitals as above. Pointing to rpmkeys in this case seems a bit dubious actually because rpmkeys wont import that package either. 

Another UI wisdom is that "allowed" tends to send the wrong kind of message, even if we think of it in those terms. Maybe just "public keys cannot be installed as packages"?

ffesti replied:
Reworded this, kept the reference to rpmkeys but made it more clear that this is not for packages.

pmatilai commented on lib/rpmte.cc:
```diff
@@ -227,6 +236,10 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
 
 exit:
+    if (rc) {
+	rpmlog(RPMLOG_ERR, _("misformed package %s\n"),
```
"misformed" is a strange word, "malformed" is the common terminology for this.

pmatilai commented on lib/rpmte.cc:
```diff
@@ -227,6 +236,10 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
+    if (rc) {
+	rpmlog(RPMLOG_ERR, _("misformed package %s\n"),
+	       p->NEVRA ? p->NEVRA : (p->name ? p->name : "-"));
+    }
```
Uff, this seems to ring a distant bell: the reason we don't have any error messages in here is that we're not guaranteed to have a meaningful handle to refer to, the only one who does is the caller of rpmtsAddInstallElement().

pmatilai replied:
So ... I think we're better off just not logging this at all. The caller gets an error code 1, if they don't report it than that's their problem, not ours.

ffesti replied:
And... it's gone.

pmatilai commented on lib/rpmte.cc:
```diff
@@ -159,9 +159,18 @@ static int addTE(rpmte p, Header h, fnpyKey key, rpmRelocation * relocs)
+    }
+
+    if (p->type != TR_REMOVED && rstreq(p->name, "gpg-pubkey")) {
+	rpmlog(RPMLOG_ERR, "public keys can not be installed as gpg-pubkey packages; use rpmkeys --import <keyfile> for that\n");
```
This and the erase warning above should be marked for translation - the _() thing around the string.

Since there's going to be at least one more round, minor nits / thoughts:
- just noticed there's a typo in one of the commit messages: "contrsucting"
- "for that" seems redundant, especially when the other message doesn't have (or need) it either
- the --delete command is something that can be copy-pasted directly, --import message isn't - maybe there should be a presentation difference (quotes around the --delete command or something)?

ffesti replied:
No, there should not. If you are seeing both messages and you get confused by them being formatted the same way, there is really nothing I can do for you.
